### ISSUE TITLE
final case classes... when they shouldn't

### ISSUE THREAD

**Issue by fommil** [scalacenter/scalafix#603]
putting `final` on an inner case class of another class give this kind of compiler warning

```
[warn] /home/fommil/Projects/scalaz-deriving/deriving-macro/src/main/scala/scalaz/deriving.scala:160:28: The outer reference in this type test cannot be checked at run time.
[warn]   final private case class TreeTypeName(tree: Tree) {
[warn]                            ^
```

so perhaps it'd be best to check if the `class` is suitable, before updating.

**Comment by olafurpg** [comment]
Can you maybe include a self-contained repro?

**Comment by fommil** [comment]
it's from this file that I seen it https://gitlab.com/fommil/scalaz-deriving/blob/master/deriving-macro/src/main/scala/scalaz/deriving.scala

I guess something like

```scala
class Foo {
  private case class Bar(s: String)
}
```

something like that.

**Comment by olafurpg** [comment]
@vovapolu any chance you can take a look at this?

**Comment by fommil** [comment]
the sponsorship period has now run out so this would be as a volunteer :sob:  (only the open PRs remain...)

**Comment by olafurpg** [comment]
Fixed in #706
